[Lauren] iP

lrnshk opened the pull request:
# Mango

> “Your mind is for having ideas, not holding them.” – David Allen

Mango frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ **SUPER FAST** to use

All you need to do is,
1. download it from [here.](https://github.com/lrnshk/ip/releases/tag/A-Jar/)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

Features:
- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

lrnshk committed:
no message

lrnshk committed:
no message

lrnshk committed:
no message

lrnshk committed:
no message

lrnshk committed:
Implemented Text Ui Testing. Fixed a bug in Todo class.

lrnshk committed:
Implemented the ability to handle errors

lrnshk committed:
Implemented the delete function. Fixed an error in Event.

lrnshk committed:
Add functionality to save and load data for the task list from the hard disk

Mango does not memorise the task list.

Mango must memorise the task list in order to track it over time.

Let's
* save the task list to hard disk
* load the task list from hard disk

lrnshk committed:
Store Deadline dates as LocalDate objects

Mango treats dates as strings and cannot interpret dates in dd-mm-yyyy format.

Users can input dates in dd-mm-yyyy format and Mango will translate them to MM dd yyyy format for readability.

Let's
* store dates as LocalDate objects
* print dates in DD mm yyyy format

lrnshk committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add functionality to save and load data for the task list from the hard disk

lrnshk committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Store Deadline dates as LocalDate objects

lrnshk committed:
Make the code more OOP

Code that are closely related are not extracted into individual classes.

Refactoring code to extract out these classes will make code more readable and flexible.

Let's
* create Parser to interpret user input
* create Ui to interact with the user
* create Storage to read and load the task list from the hard disk
* create TaskList to house the task list and perform operations on the task list

lrnshk committed:
Organise the classes into suitable java packages

All classes are in unnamed packages.

Grouping the classes into their respective named packages makes the program code more organised.

Let's
* Group related classes into their own package
* Group all classes into the mango package

lrnshk committed:
Add JUnit tests to test the behavior of the task package

Code has no test that checks if it works correctly.

JUnit automated tasks can check if the methods in the various classes such as Deadline in the task package are working as they should be.

Let's
* implement some JUnit tests in a TaskTest class
* test the getDescription() and isValid() methods

lrnshk committed:
Add JavaDoc comments to most of the non-private classes and methods.

The code had no comments.

With comments, the code is more readable.

Let's
* Add comments to the non-private classes and methods

lrnshk committed:
Change some code to fulfil Java coding standard.

Some of the code did not follow Java coding standard.

Following the coding standard makes code neat and readable.

Let's
* Edit the code to follow the Java coding standard.

lrnshk committed:
Add the functionality to search for a keyword in the task list.

No search function available.

User can easily find the task they are looking for by inputting a keyword.

Let's
* Tweak the Parser to accept the "find" command
* Include a find() method for the task list to locate tasks that contain the keyword.

lrnshk committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments to most of the non-private classes and methods.

lrnshk committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Change some code to fulfil Java coding standard.

# Conflicts:
#	src/main/java/mango/TaskList.java

lrnshk committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add the functionality to search for a keyword in the task list.

# Conflicts:
#	src/main/java/mango/TaskList.java

lrnshk committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d'

* commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d':
  Add Gradle support

lrnshk committed:
Use Gradle to automate some of the build tasks of the project.

lrnshk committed:
Add a GUI to Mango using the JavaFX technology.

Mango uses a text UI.

A GUI is more user-friendly and aesthetically pleasing.

Let's
* Add a GUI to Mango

lrnshk committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add a GUI to Mango using the JavaFX technology.

EthanWong6362 commented on src/main/java/mango/Storage.java:
```diff
@@ -0,0 +1,95 @@
+package mango;
+
+import mango.task.*;
```
Imported classes should always be listed explicitly instead of using * import

EthanWong6362 commented on src/main/java/mango/Duke.java:
```diff
@@ -0,0 +1,76 @@
+     */
+    public String exit() {
+        try {
+            this.storage.save(this.tasks);
```
Honestly, I just really like this method of saving to file at the end instead of constantly, nice!

sivayogasubramanian replied:
This might be a problem if the application crashes (due to an exception) while the user is using it. All the tasks created then will not be saved. There is a forum post on this.

EthanWong6362 commented on src/main/java/mango/task/Deadline.java:
```diff
@@ -0,0 +1,79 @@
+        this.date = LocalDate.parse(date);
+    }
+
+    /**
+     * Returns the string representation of the type of task.
+     *
+     * @return The string representation of the type of task.
+     */
+    public String getType() {
+        return "D";
+    }
```
I feel like these methods should be `@Override`d for sake of clarity?

EthanWong6362 reviewed (commented):
Really clean, easily readable code. Good work!!

sivayogasubramanian commented on build.gradle:
```diff
@@ -0,0 +1,62 @@
+    toolVersion = '8.29'
+}
+
+run{
```
Add space here
```suggestion
run {
```

sivayogasubramanian commented on src/main/java/mango/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.Collections;
+
+/**
+ * An example of a custom control using FXML.
```
This line should probably be removed

sivayogasubramanian commented on src/main/java/mango/MainWindow.java:
```diff
@@ -0,0 +1,56 @@
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
+    }
+
+    public void setDuke(Duke d) {
```
Javadoc is missing for this `public` method

sivayogasubramanian commented on src/main/java/mango/MainWindow.java:
```diff
@@ -0,0 +1,56 @@
+    private Image dukeImage = new Image(this.getClass().getResourceAsStream("/images/Mango.jpg"));
+
+    @FXML
+    public void initialize() {
```
Javadoc is missing for this `public` method

sivayogasubramanian commented on src/main/java/mango/TaskList.java:
```diff
@@ -0,0 +1,184 @@
+package mango;
+
+import mango.task.*;
```
Imported classes should always be listed explicitly instead of using * import

sivayogasubramanian commented on src/main/java/mango/TaskList.java:
```diff
@@ -0,0 +1,184 @@
+     * @return The confirmation message that the task has been added.
+     * @throws DukeException If the input string does not make sense.
+     */
+    public String add(String str) throws DukeException {
```
```suggestion
    public String addTask(String str) throws DukeException {
```

sivayogasubramanian commented on src/main/java/mango/TaskList.java:
```diff
@@ -0,0 +1,184 @@
+     * @param completedTask The task that is to be marked as done.
+     * @return The confirmation message that the task has been marked completed.
+     */
+    public String complete(int completedTask) {
```
```suggestion
    public String completeTask(int completedTaskIndex) {
```

sivayogasubramanian commented on src/main/java/mango/TaskList.java:
```diff
@@ -0,0 +1,184 @@
+     * @param deleteTask The task to be deleted.
+     * @return The confirmation message that the task has been deleted.
+     */
+    public String delete(int deleteTask) {
```
```suggestion
    public String deleteTask(int deleteTaskIndex) {
```

sivayogasubramanian commented on src/main/java/mango/TaskList.java:
```diff
@@ -0,0 +1,184 @@
+     * @param str The string to find.
+     * @return The tasks that match the keyword.
+     */
+    public String find(String str) {
```
```suggestion
    public String findTask(String str) {
```

sivayogasubramanian commented on src/main/java/mango/TaskList.java:
```diff
@@ -0,0 +1,184 @@
+     *
+     * @return All tasks in the task list in a string.
+     */
+    public String printList() {
```
This can just be `toString()` right?

sivayogasubramanian commented on src/main/java/mango/task/Event.java:
```diff
@@ -0,0 +1,76 @@
+     *
+     * @return The string representation of the event.
+     */
+    public String save() {
```
The name does not really tell that it returns a `String`. Sounds like an action where i am saving something.
```suggestion
    public String getSaveFormatString() {
```

sivayogasubramanian commented on src/main/resources/view/DialogBox.fxml:
```diff
@@ -0,0 +1,16 @@
+<?import javafx.scene.image.ImageView?>
+<?import javafx.scene.layout.HBox?>
+
+<fx:root alignment="TOP_RIGHT" maxHeight="1.7976931348623157E308" maxWidth="1.7976931348623157E308" prefWidth="400.0" type="javafx.scene.layout.HBox" xmlns="http://javafx.com/javafx/8.0.171" xmlns:fx="http://javafx.com/fxml/1">
```
This line is too long

sivayogasubramanian commented on src/main/resources/view/MainWindow.fxml:
```diff
@@ -0,0 +1,23 @@
+<?import javafx.scene.layout.VBox?>
+<?import javafx.scene.text.Font?>
+
+<AnchorPane maxHeight="-Infinity" maxWidth="-Infinity" minHeight="-Infinity" minWidth="-Infinity" prefHeight="600.0" prefWidth="400.0" xmlns="http://javafx.com/javafx/16" xmlns:fx="http://javafx.com/fxml/1" fx:controller="mango.MainWindow">
```
This line is too long

sivayogasubramanian commented on src/main/resources/view/MainWindow.fxml:
```diff
@@ -0,0 +1,23 @@
+
+<AnchorPane maxHeight="-Infinity" maxWidth="-Infinity" minHeight="-Infinity" minWidth="-Infinity" prefHeight="600.0" prefWidth="400.0" xmlns="http://javafx.com/javafx/16" xmlns:fx="http://javafx.com/fxml/1" fx:controller="mango.MainWindow">
+    <children>
+        <TextField fx:id="userInput" layoutY="558.0" onAction="#handleUserInput" prefHeight="38.0" prefWidth="324.0" AnchorPane.bottomAnchor="1.0" AnchorPane.leftAnchor="2.0" />
+        <Button fx:id="sendButton" layoutX="324.0" layoutY="561.0" mnemonicParsing="false" onAction="#handleUserInput" prefHeight="38.0" prefWidth="69.0" text="send" textFill="#9d00ff" AnchorPane.rightAnchor="2.0">
```
Lines are too long

sivayogasubramanian reviewed (commented):
I added some non-coding violation comments too! Hope you don't mind. There there very little coding violations. :)

lrnshk committed:
Add Assertions to the code

Assert feature not used to check for assumptions.

Assert features will enable runtime to verify assumptions in the code, which helps identify faults.

Let's
* Add assertions to parts of the code that make assumptions about certain variables.

lrnshk committed:
Improve code quality

The code was not up to the code quality standards.

Refactoring the code according to the code quality standards makes it cleaner and more readable.

Let's
* Refactor the code according to the code quality standards

lrnshk committed:
Merge pull request #1 from lrnshk/branch-A-Assertions

Add Assertions to the code

lrnshk committed:
Merge branch 'master' of https://github.com/lrnshk/ip

* 'master' of https://github.com/lrnshk/ip:
  Add Assertions to the code

lrnshk committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add Assertions to the code

# Conflicts:
#	src/main/java/mango/TaskList.java

lrnshk committed:
Merge pull request #2 from lrnshk/branch-A-CodeQuality

Improve code quality

lrnshk committed:
Merge branch 'master' of https://github.com/lrnshk/ip

* 'master' of https://github.com/lrnshk/ip:
  Improve code quality

lrnshk committed:
Provide functionality to tag tasks with strings

There is no option to tag tasks.

Tagging tasks would enable users to specify what kind of task they have to do and possibly group them based on these tags.

Let's
* Add some functionality to tag tasks with strings.

lrnshk committed:
Merge branch 'branch-C-Tagging'

* branch-C-Tagging:
  Provide functionality to tag tasks with strings

lrnshk committed:
Improve the GUI.

GUI is not aesthetically attractive and misaligned.

A better GUI will look more organised and appealing to use.

Let's
* Add background colours to the main window and dialog boxes
* Add padding to align the text
* Fix the height of the text boxes so that it is not compressed during overflow
* Reduce the size of user dialog box relative to Mango's to save space

lrnshk committed:
Add a User Guide for Mango

Mango has no user guide.

A User Guide will be helpful to a new user by providing information on what commands are available.

Let's
* Add a User Guide in README.md

lrnshk committed:
Fixed a bug in TaskTest
